_posts/2019-06-25-dynamic_linear_models_tfprobability/dynamic_linear_models_tfprobability.Rmd

```diff
@@ -0,0 +1,805 @@
+      ts_train %>% sts_build_factored_variational_loss(model = model)
+    variational_loss <- loss_and_dists[[1]]
+    train_op <- optimizer$minimize(variational_loss)
+    
+    with (tf$Session() %as% sess,  {
+      
+      sess$run(tf$compat$v1$global_variables_initializer())
+      for (step in 1:n_iterations) {
+        sess$run(train_op)
+        loss <- sess$run(variational_loss)
+        if (step %% 1 == 0)
+          cat("Loss: ", as.numeric(loss), "\n")
+      }
+      variational_distributions <- loss_and_dists[[2]]
+      posterior_samples <-
+        Map(
+          function(d)
+            d %>% tfd_sample(n_param_samples),
+          variational_distributions %>% reticulate::py_to_r() %>% unname()
+        )
+      forecast_dists <-
+        ts_train %>% sts_forecast(model, posterior_samples, n_forecast_steps)
+      fc_means <- forecast_dists %>% tfd_mean()
+      fc_sds <- forecast_dists %>% tfd_stddev()
+      
+      ssm <- model$make_state_space_model(length(ts_train), param_vals = posterior_samples)
+      c(smoothed_means, smoothed_covs) %<-% ssm$posterior_marginals(ts_train)
+      c(., filtered_means, filtered_covs, ., ., ., .) %<-% ssm$forward_filter(ts_train)
+   
+      c(posterior_samples, fc_means, fc_sds, smoothed_means, smoothed_covs, filtered_means, filtered_covs) %<-%
+        sess$run(list(posterior_samples, fc_means, fc_sds, smoothed_means, smoothed_covs, filtered_means, filtered_covs))
+      
+    })
+    
+    list(
+      variational_distributions,
+      posterior_samples,
+      fc_means[, 1],
+      fc_sds[, 1],
+      smoothed_means,
+      smoothed_covs,
+      filtered_means,
+      filtered_covs
+    )
+  }
+```
+
+
+And this is how we call it. 
+
+```{r}
+# number of VI steps
+n_iterations <- 300
+# sample size for posterior samples
+n_param_samples <- 50
+# sample size to draw from the forecast distribution
+n_forecast_samples <- 50
+
+# here's the model again
+model <- ts %>%
+  sts_dynamic_linear_regression(design_matrix = cbind(rep(1, length(x)), x) %>% tf$cast(tf$float32))
+
+# call fit_vi defined above
+c(
+  param_distributions,
+  param_samples,
+  fc_means,
+  fc_sds,
+  smoothed_means,
+  smoothed_covs,
+  filtered_means,
+  filtered_covs
+) %<-% fit_vi(
+  ts,
+  ts_train,
+  model,
+  n_iterations,
+  n_param_samples,
+  n_forecast_steps,
+  n_forecast_samples
+)
+
+```
+
+Curious about the results? We'll see them in a second, but before let's just quickly glance at the alternative training method: HMC.
+
+#### Putting it all together (the HMC edition)
+
+`tfprobability` provides [sts_fit_with_hmc](https://rstudio.github.io/tfprobability/reference/sts_fit_with_hmc.html) to fit a DLM using Hamiltonian Monte Carlo. Recent posts (e.g., [Hierarchical partial pooling, continued: Varying slopes models with TensorFlow Probability](https://blogs.rstudio.com/tensorflow/posts/2019-05-24-varying-slopes/)) showed how to set up HMC to fit hierarchical models; here a single function does it all.
```

Comment:
Here you're not using parens after the function name either. I think it works both ways, but you want to be consistent.